fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -3531,40 +3571,70 @@ public Subspace getLegacyVersionSubspace() {
+                    lazyRecordCount.get().thenApply(recordCount -> FDBRecordStore.disabledIfTooManyRecordsForRebuild(recordCount, indexOnNewRecordTypes)) :
+                    userVersionChecker.needRebuildIndex(index, lazyRecordCount, lazyRecordsSize, indexOnNewRecordTypes);
             if (index.getType().equals(IndexTypes.VERSION)
                     && !newStore
                     && oldFormatVersion < SAVE_VERSION_WITH_RECORD_FORMAT_VERSION
-                    && !useOldVersionFormat()
-                    && state.equals(IndexState.READABLE)) {
-                // Do not rebuild any version indexes while the format conversion is going on.
-                // Otherwise, the process moving the versions might race against the index
-                // build and some versions won't be indexed correctly.
-                state = IndexState.DISABLED;
+                    && !useOldVersionFormat()) {
+                stateFuture = stateFuture.thenApply(state -> {
+                    if (IndexState.READABLE.equals(state)) {
+                        // Do not rebuild any version indexes while the format conversion is going on.
+                        // Otherwise, the process moving the versions might race against the index
+                        // build and some versions won't be indexed correctly.
+                        return IndexState.DISABLED;
+                    }
+                    return state;
+                });
             }
-            newStates.put(index, state);
+            newStates.put(index, stateFuture);
         }
+        return newStates;
+    }
+
+    private void maybeLogIndexesNeedingRebuilding(@Nonnull Map<Index, CompletableFuture<IndexState>> newStates,
+                                                  @Nonnull AtomicLong recordCountRef,
+                                                  @Nonnull AtomicLong recordsSizeRef,
+                                                  boolean rebuildRecordCounts,
+                                                  boolean newStore) {
         if (LOGGER.isDebugEnabled()) {
             KeyValueLogMessage msg = KeyValueLogMessage.build("indexes need rebuilding",
-                                        LogMessageKeys.RECORD_COUNT, recordCount == Long.MAX_VALUE ? "unknown" : Long.toString(recordCount), 
-                                        subspaceProvider.logKey(), subspaceProvider.toString(context));
+                    subspaceProvider.logKey(), subspaceProvider.toString(context));
+
+            long recordCount = recordCountRef.get();
+            if (recordCount >= 0L) {
+                msg.addKeyAndValue(LogMessageKeys.RECORD_COUNT, recordCount == Long.MAX_VALUE ? "unknown" : Long.toString(recordCount));
+            }
+            long recordsSize = recordsSizeRef.get();
+            if (recordsSize >= 0L) {
+                msg.addKeyAndValue(LogMessageKeys.RECORDS_SIZE_ESTIMATE, Long.toString(recordsSize));
+            }
```

Comment:
I think this deserves a comment: if the checker called the supplier, then log what it yielded.